### PR TITLE
Adapts webapp for expansion of giftable project types

### PR DESCRIPTION
**Pull request by mohitb35**
1. Adapts bulk code logic to allow gifts to other projects besides those with `purpose: trees`
2. Adapts payment history logic to show deliverables (downloads) wherever applicable. `bouquet` and `composite` projects will not show Donor Receipt, Gift Receipt, or have downloadable codes, but they can have Tax Receipts. For other projects, deliverables are shown based on the API response.
3. Adapts logic/ui for successful redemption message
4. Refactors to resolve typescript warnings in files used for the redeem code, bulk codes, payment history features, and fixes minor typos
5. Updates types to reflect updated API responses while redeeming codes, creating bulk codes
6. Adds validation to check for long recipient names while creating discrete bulk codes, both while uploading an excel/csv,  or while adding individual recipients using the form
7. Updates projects API to show other projects for selection while generating bulk codes (besides tree projects). `fund` projects are currently disabled by using `filter[purpose]`, but the logic to allow fund projects is in place. This includes updates to types (`ProjectMinimal` instead of `ProjectOption`), logic to filter allowed projects from API response, changes to logic for `BulkGiftTotal` and `UnitCostDisplay`

Pending:

- [x] Update /projects API to show all selection of all applicable projects that permit bulk code generation. Currently fund projects are disabled.
- [x] Only allow selection of projects that are supported by the country entity (based on planetcash account country) - requires API updates.

### CODE REVIEW

File: src/features/user/BulkCodes/components/BulkGiftTotal.tsx [hunk 1]
```diff
@@ -21,25 +23,31 @@ const BulkGiftTotal = ({
...
-      // TODOO translation and pluralization
+      value={tBulkCodes('summaryTotal', {
+        formattedAmount: getFormattedCurrency(
+          locale,
+          currency as string,
+          amount
+        ),
+        units,
+        pluralizedUnitType: getPluralizedUnitType(unit, units),
+      })}
```
[the remark this answers is not in the document]
Comment (mohitb35) [reply src/features/user/BulkCodes/components/BulkGiftTotal.tsx]: 79255d154cd51c671d490d0aa2abf904999277bd resolves this

File: src/features/common/RedeemCode/SuccessfullyRedeemed.tsx [hunk 2]
```diff
@@ -29,17 +27,16 @@ export const SuccessfullyRedeemed = ({
...
+      <div className={styles.successMessage}>
+        {redeemedCodeData?.project.classification === 'membership'
+          ? tRedeem.rich('membershipRedeemSuccessMessage', {
+              line1: (chunks) => <p>{chunks}</p>,
+              line2: (chunks) => <p>{chunks}</p>,
+            })
+          : tRedeem.rich('redeemSuccessMessage', {
+              line1: (chunks) => <p>{chunks}</p>,
+              line2: (chunks) => <p>{chunks}</p>,
+            })}
```
[the remark this answers is not in the document]
Comment (mohitb35) [reply src/features/common/RedeemCode/SuccessfullyRedeemed.tsx]: 79255d154cd51c671d490d0aa2abf904999277bd resolves this

File: src/features/common/types/project.d.ts [hunk 3]
```diff
@@ -269,11 +269,13 @@ export interface ProjectOption {
   name: string;
   unitCost: number;
   currency: string;
-  unit?: 'tree' | 'm2' | 'ha';
-  purpose: string;
+  unit?: UnitType;
+  purpose: 'trees' | 'conservation' | 'funds';
```
[the remark this answers is not in the document]
Comment (mohitb35) [reply src/features/common/types/project.d.ts:273]: This is not a concern. In face, `bouquet` and `composite` projects are not applicable in places this type is used.

File: src/features/common/RedeemCode/SuccessfullyRedeemed.tsx [hunk 4]
```diff
@@ -29,17 +27,16 @@ export const SuccessfullyRedeemed = ({
...
+      <div className={styles.successMessage}>
+        {redeemedCodeData?.project?.classification === 'membership'
+          ? tRedeem.rich('membershipRedeemSuccessMessage', {
+              line1: (chunks) => <p>{chunks}</p>,
+              line2: (chunks) => <p>{chunks}</p>,
+            })
+          : tRedeem.rich('redeemSuccessMessage', {
+              line1: (chunks) => <p>{chunks}</p>,
+              line2: (chunks) => <p>{chunks}</p>,
+            })}
```
[the remark this answers is not in the document]
Comment (mohitb35) [reply src/features/common/RedeemCode/SuccessfullyRedeemed.tsx:39]: This isn't needed.

File: src/features/user/BulkCodes/BulkCodesTypes.d.ts [hunk 5]
```diff
@@ -1,15 +1,35 @@
...
-  unit: 'tree' | 'm2' | 'ha';
+  unitType: UnitTypes;
 }
 
+export interface TreePaymentOptions extends PaymentOptionsBase {
+  purpose: 'trees';
+  unitType: 'tree' | 'm2';
+}
+
+export interface ConservationPaymentOptions extends PaymentOptionsBase {
```
[the remark this answers is not in the document]
Comment (mohitb35) [reply src/features/user/BulkCodes/BulkCodesTypes.d.ts:18]: Keeping it in PaymentOptionsBase ensures that the types that extend PaymentOptionsBase do not contain an invalid value for `unitType`. For example, if I defined `unitType: 'kg'`, I'd see a warning.

**Review by sunilsabatp — APPROVED** [review]

**Review by mariahosfeld — COMMENTED** [review]
Functionality looks good to me.

**Pull request merged**